File: source/backbone/lib/packet/packet.cpp
#include "packet.hpp"

/* ------------------------------------------------------------------------------------------------------- */

Packet::Packet(PacketHeader                header,
               std::vector<PacketQuestion> questions,
               std::vector<PacketRecord>   answers,
               std::vector<PacketRecord>   authorities,
               std::vector<PacketRecord>   additionals)
    : header(header), questions(questions), answers(answers), authorities(authorities),
      additionals(additionals) {}

/* ------------------------------------------------------------------------------------------------------- */

Result<Packet>
Packet::from_buffer(Ref<PacketBuffer> buffer) {
   /* Parse header */
   auto header = PacketHeader::from_buffer(buffer).except("Failed to parse packet header");
   RETURN_IF_ERROR(header);

   /* Parse questions */
   std::vector<PacketQuestion> questions;
   for (size_t i = 0; i < header.get_value().question_count; i++) {
      // Parse question
      auto question = PacketQuestion::from_buffer(buffer).except("Failed to parse packet question");
      RETURN_IF_ERROR(question);

      // Add question to the list
      questions.push_back(question.get_value());
   }

   /* Parse answers */
   std::vector<PacketRecord> answers;
   for (size_t i = 0; i < header.get_value().answer_count; i++) {
      // Parse answer
      auto answer = PacketRecord::from_buffer(buffer).except("Failed to parse packet answer");
      RETURN_IF_ERROR(answer);

      // Add answer to the list
      answers.push_back(answer.get_value());
   }

   /* Parse authorities */
   std::vector<PacketRecord> authorities;
   for (size_t i = 0; i < header.get_value().authority_count; i++) {
      // Parse authority
      auto authority = PacketRecord::from_buffer(buffer).except("Failed to parse packet authority");
      RETURN_IF_ERROR(authority);

      // Add authority to the list
      authorities.push_back(authority.get_value());
   }

   /* Parse additionals */
   std::vector<PacketRecord> additionals;
   for (size_t i = 0; i < header.get_value().additional_count; i++) {
      // Parse additional
      auto additional = PacketRecord::from_buffer(buffer).except("Failed to parse packet additional");
      RETURN_IF_ERROR(additional);

      // Add additional to the list
      additionals.push_back(additional.get_value());
   }

   auto packet = Packet(header.get_value(), questions, answers, authorities, additionals);
   return packet;
}

/* ------------------------------------------------------------------------------------------------------- */

Result<void>
Packet::write_to_buffer(Ref<PacketBuffer> buffer) const {
   auto res = Ok();

   /* Write header */
   res = this->header.write_to_buffer(buffer).except("Failed to write packet header");
   RETURN_IF_ERROR(res);

   /* Write questions */
   for (const auto &question : questions) {
      res = question.write_to_buffer(buffer).except("Failed to write packet question");
      RETURN_IF_ERROR(res);
   }

   /* Write answers */
   for (const auto &answer : answers) {
      res = answer.write_to_buffer(buffer).except("Failed to write packet answer");
      RETURN_IF_ERROR(res);
   }

   /* Write authorities */
   for (const auto &authority : authorities) {
      res = authority.write_to_buffer(buffer).except("Failed to write packet authority");
      RETURN_IF_ERROR(res);
   }

   /* Write additionals */
   for (const auto &additional : additionals) {
      res = additional.write_to_buffer(buffer).except("Failed to write packet additional");
      RETURN_IF_ERROR(res);
   }

   return Ok();
}

/* ------------------------------------------------------------------------------------------------------- */

File: src/dns/packet/packet.rs
use crate::{
    dns::{question::question::DNSQuestion, record::record::DNSRecord},
    utility::result::Result,
};

use super::{buffer::PacketBufferOps, header::PacketHeader};

#[derive(Clone, Debug)]
pub struct Packet {
    pub header: PacketHeader,
    pub questions: Vec<DNSQuestion>,
    pub answers: Vec<DNSRecord>,
    pub authorities: Vec<DNSRecord>,
    pub additionals: Vec<DNSRecord>,
}

impl Packet {
    pub fn new() -> Packet {
        Packet {
            header: PacketHeader::new(),
            questions: Vec::new(),
            answers: Vec::new(),
            authorities: Vec::new(),
            additionals: Vec::new(),
        }
    }

    pub fn from_buffer<T: PacketBufferOps>(buffer: &mut T) -> Result<Packet> {
        let mut result = Packet::new();

        /* Read header */
        result.header = PacketHeader::from_buffer(buffer)?;

        /* Read questions */
        for _ in 0..result.header.question_count {
            DNSQuestion::from_buffer(buffer)
                .and_then(|question| {
                    result.questions.push(question);
                    Ok(())
                })
                .expect("Failed to read question from buffer.");
        }

        /* Read answers */
        for _ in 0..result.header.answer_count {
            DNSRecord::from_buffer(buffer)
                .and_then(|record| {
                    result.answers.push(record);
                    Ok(())
                })
                .expect("Failed to read answer from buffer.");
        }

        /* Read authorities */
        for _ in 0..result.header.authority_count {
            DNSRecord::from_buffer(buffer)
                .and_then(|record| {
                    result.authorities.push(record);
                    Ok(())
                })
                .expect("Failed to read authority from buffer.");
        }

        /* Read additionals */
        for _ in 0..result.header.additional_count {
            DNSRecord::from_buffer(buffer)
                .and_then(|record| {
                    result.additionals.push(record);
                    Ok(())
                })
                .expect("Failed to read additional from buffer.");
        }

        Ok(result)
    }
}
